### ISSUE TITLE
12$: Error on invalid syntax in object types

### ISSUE THREAD

**Issue by strager** [quick-lint/quick-lint-js#765]
```typescript
// generator * or async not allowed on methods
type O = { *method() }
type O = { async method() }

// + and - on readonly and ? not allowed on computed-named or normal properties
type O = { +readonly [key]: typexpr, }
type O = { [key]+?: typexpr, }
type O = { prop+?: typexpr, }
type O = { +readonly prop: typexpr, }

// index signatures cannot be methods
type O = { [key in typeexpr]() }
type O = { [key: typeexpr]() }

// index signatures must have a field type
type O = { [key: typeexpr] }

// index signatures cannot be optional
type O = { [key: typeexpr]?: typexpr, }
type O = { [key: typeexpr]+?: typexpr, }

// getters and setters must be methods
type O = { get name: Type }
type O = { set name: Type }

// methods cannot have bodies
type O = { method() {} }

// methods cannot be readonly
type O = { readonly method() }

// setters cannot have return types
type O = { set name(v): Type }

// index signature variables cannot be named 'let' or 'static'
type O = { [let: Type]: Type }
type O = { [static: Type]: Type }
```
